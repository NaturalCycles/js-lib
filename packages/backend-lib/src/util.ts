const { GAE_INSTANCE, K_SERVICE } = process.env

export function isGAE(): boolean {
  return !!GAE_INSTANCE
}

export function isCloudRun(): boolean {
  return !!K_SERVICE
}
