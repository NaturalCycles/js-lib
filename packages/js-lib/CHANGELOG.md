See [Github Releases](https://github.com/NaturalCycles/js-lib/releases)
