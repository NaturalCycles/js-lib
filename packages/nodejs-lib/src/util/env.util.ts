import 'dotenv/config' // ensure .env is read before requiring keys
import type { ValuesOf } from '@naturalcycles/js-lib'
import { fs2 } from '../fs/fs2.js'

/**
 * @example
 *
 * const {a, b} = requreEnvKeys(['a', 'b'])
 *
 * Will throw if any of the passed keys is not defined.
 */
export function requireEnvKeys<T extends readonly string[]>(
  ...keys: T
): { [k in ValuesOf<T>]: string } {
  // eslint-disable-next-line unicorn/no-array-reduce
  return keys.reduce(
    (r, k) => {
      const v = process.env[k]
      if (!v) throw new Error(`${k} env variable is required, but missing`)
      r[k as ValuesOf<T>] = v
      return r
    },
    {} as { [k in ValuesOf<T>]: string },
  )
}

/**
 * @deprecated use fs2.requireFileToExist
 */
export function requireFileToExist(filePath: string): void {
  fs2.requireFileToExist(filePath)
}
