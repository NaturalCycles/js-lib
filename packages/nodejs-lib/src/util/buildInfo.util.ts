import type { AnyObject, BuildInfo, UnixTimestamp } from '@naturalcycles/js-lib'
import { _filterUndefinedValues, localTime } from '@naturalcycles/js-lib'
import { fs2 } from '../fs/fs2.js'
import { git2 } from './git2.js'

export interface GenerateBuildInfoOptions {
  /**
   * If set - this timestamp will be used, instead of "current time".
   */
  overrideTimestamp?: UnixTimestamp

  /**
   * Defaults to currently checked out git branch.
   */
  overrideBranchName?: string
}

export function generateBuildInfo(opt: GenerateBuildInfoOptions = {}): BuildInfo {
  const now = localTime.orNow(opt.overrideTimestamp)
  const ts = now.unix

  const rev = git2.getCurrentCommitSha()
  const branchName = opt.overrideBranchName || git2.getCurrentBranchName()
  const repoName = git2.getCurrentRepoName()
  const tsCommit = git2.getCurrentCommitTimestamp()

  const ver = [now.toStringCompact(), repoName, branchName, rev].join('_')

  let { APP_ENV: env } = process.env

  if (!env) {
    // Attempt to read `envByBranch` from package.json root
    try {
      if (fs2.pathExists('package.json')) {
        const packageJson = fs2.readJson<AnyObject>('package.json')
        env = packageJson?.['envByBranch']?.[branchName] || packageJson?.['envByBranch']?.['*']
      }
    } catch {}
  }

  return _filterUndefinedValues({
    ts,
    tsCommit,
    repoName,
    branchName,
    rev,
    ver,
    env,
  })
}
