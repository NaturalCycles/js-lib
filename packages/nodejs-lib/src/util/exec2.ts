import { execSync, spawn, spawnSync } from 'node:child_process'
import {
  _substringAfterLast,
  type AnyObject,
  type NumberOfMilliseconds,
  type UnixTimestampMillis,
} from '@naturalcycles/js-lib'
import { _since, AppError } from '@naturalcycles/js-lib'
import { dimGrey, dimRed, hasColors, white } from '../colors/colors.js'

/**
 * Set of utility functions to work with Spawn / Exec.
 *
 * How to decide between Spawn and Exec?
 *
 * Long-running job that prints output, and no need to return the output - use Spawn.
 *
 * Short-running job, no need to print the output, might want to return the output - use Exec.
 *
 * Need to both print and return the output - use SpawnAsyncAndReturn.
 *
 * ***
 *
 * Spawn is good for long-running large-output processes, that continuously output data.
 * E.g running `jest`.
 *
 * Exec is the opposite - good for short-running processes that output small data.
 * Exec allows to return the output as a string.
 * Exec doesn't stream data during execution, so the output/error will only be printed
 * at the end.
 * Exec always uses the shell (there's no option to disable it).
 */
class Exec2 {
  /**
   * Reasons to use it:
   * - Sync
   * - Need to print output while running
   *
   * Limitations:
   * - Cannot return stdout/stderr (use exec, execAsync or spawnAsyncAndReturn for that)
   *
   * Defaults:
   *
   * shell: true
   * log: true
   */
  spawn(cmd: string, opt: SpawnOptions = {}): void {
    const { shell = true, cwd, env, passProcessEnv = true, forceColor = hasColors } = opt
    opt.log ??= true // by default log should be true, as we are printing the output
    opt.logStart ??= opt.log
    opt.logFinish ??= opt.log
    const started = Date.now() as UnixTimestampMillis
    this.logStart(cmd, opt)

    const r = spawnSync(cmd, opt.args, {
      encoding: 'utf8',
      stdio: 'inherit',
      shell,
      cwd,
      env: {
        ...(passProcessEnv ? process.env : {}),
        ...(forceColor ? { FORCE_COLOR: '1' } : {}),
        ...env,
      },
    })

    const isSuccessful = !r.error && !r.status
    this.logFinish(cmd, opt, started, isSuccessful)

    if (r.error) {
      throw r.error
    }
    if (r.status) {
      throw new Error(`spawn exited with code ${r.status}: ${cmd}`)
    }
  }

  /**
   * Reasons to use it:
   *
   * - Sync
   * - Need to return output
   *
   * Limitations:
   * - Cannot print while running (use spawn or spawnAsync for that)
   *
   * Defaults:
   *
   * shell: true
   * log: false
   */
  exec(cmd: string, opt: ExecOptions = {}): string {
    const { cwd, env, passProcessEnv = true, timeout } = opt
    opt.logStart ??= opt.log ?? false
    opt.logFinish ??= opt.log ?? false
    const started = Date.now() as UnixTimestampMillis
    this.logStart(cmd, opt)

    try {
      const s = execSync(cmd, {
        encoding: 'utf8',
        // stdio: 'inherit', // no, otherwise we don't get the output returned
        stdio: undefined,
        // shell: undefined,
        cwd,
        timeout,
        env: {
          ...(passProcessEnv ? process.env : {}),
          ...env,
        },
      }).trim()

      this.logFinish(cmd, opt, started, true)
      return s
    } catch (err) {
      // Not logging stderr, as it's printed by execSync by default (somehow)
      // stdout is not printed by execSync though, therefor we print it here
      // if ((err as any).stderr) {
      //   process.stderr.write((err as any).stderr)
      // }
      if ((err as any).stdout) {
        process.stdout.write((err as any).stdout)
      }
      this.logFinish(cmd, opt, started, false)
      throw new Error(`exec exited with code ${(err as any).status}: ${cmd}`)
    }
  }

  /**
   * Reasons to use it:
   * - Async
   * - Need to print output while running
   *
   * Limitations:
   * - Cannot return stdout/stderr (use execAsync or spawnAsyncAndReturn for that)
   *
   * Defaults:
   *
   * shell: true
   * log: true
   */
  async spawnAsync(cmd: string, opt: SpawnOptions = {}): Promise<void> {
    const { shell = true, cwd, env, passProcessEnv = true, forceColor = hasColors } = opt
    opt.log ??= true // by default log should be true, as we are printing the output
    opt.logStart ??= opt.log
    opt.logFinish ??= opt.log
    const started = Date.now() as UnixTimestampMillis
    this.logStart(cmd, opt)

    await new Promise<void>((resolve, reject) => {
      const p = spawn(cmd, opt.args || [], {
        shell,
        cwd,
        stdio: 'inherit',
        env: {
          ...(passProcessEnv ? process.env : {}),
          ...(forceColor ? { FORCE_COLOR: '1' } : {}),
          ...env,
        },
      })

      p.on('close', code => {
        const isSuccessful = !code
        this.logFinish(cmd, opt, started, isSuccessful)
        if (code) {
          return reject(new Error(`spawnAsync exited with code ${code}: ${cmd}`))
        }
        resolve()
      })
    })
  }

  /**
   * Advanced/async version of Spawn.
   * Consider simpler `spawn` or `exec` first, which are also sync.
   *
   * spawnAsyncAndReturn features:
   *
   * 1. Async
   * 2. Allows to collect the output AND print it while running.
   * 3. Returns SpawnOutput with stdout, stderr and exitCode.
   * 4. Allows to not throw on error, but just return SpawnOutput for further inspection.
   *
   * Defaults:
   *
   * shell: true
   * printWhileRunning: true
   * collectOutputWhileRunning: true
   * throwOnNonZeroCode: true
   * log: true
   */
  async spawnAsyncAndReturn(cmd: string, opt: SpawnAsyncOptions = {}): Promise<SpawnOutput> {
    const {
      shell = true,
      printWhileRunning = true,
      collectOutputWhileRunning = true,
      throwOnNonZeroCode = true,
      cwd,
      env,
      passProcessEnv = true,
      forceColor = hasColors,
    } = opt
    opt.log ??= printWhileRunning // by default log should be true, as we are printing the output
    opt.logStart ??= opt.log
    opt.logFinish ??= opt.log
    const started = Date.now() as UnixTimestampMillis
    this.logStart(cmd, opt)
    let stdout = ''
    let stderr = ''

    return await new Promise<SpawnOutput>((resolve, reject) => {
      const p = spawn(cmd, opt.args || [], {
        shell,
        cwd,
        env: {
          ...(passProcessEnv ? process.env : {}),
          ...(forceColor ? { FORCE_COLOR: '1' } : {}),
          ...env,
        },
      })

      p.stdout.on('data', data => {
        if (collectOutputWhileRunning) {
          stdout += data.toString()
          // console.log('stdout:', data.toString())
        }
        if (printWhileRunning) {
          process.stdout.write(data)
          // console.log('stderr:', data.toString())
        }
      })
      p.stderr.on('data', data => {
        if (collectOutputWhileRunning) {
          stderr += data.toString()
        }
        if (printWhileRunning) {
          process.stderr.write(data)
        }
      })

      p.on('close', code => {
        const isSuccessful = !code
        this.logFinish(cmd, opt, started, isSuccessful)
        const exitCode = code || 0
        const o: SpawnOutput = {
          exitCode,
          stdout: stdout.trim(),
          stderr: stderr.trim(),
        }
        if (throwOnNonZeroCode && code) {
          return reject(new SpawnError(`spawnAsyncAndReturn exited with code ${code}: ${cmd}`, o))
        }
        resolve(o)
      })
    })
  }

  private logStart(cmd: string, opt: SpawnOptions | ExecOptions): void {
    if (!opt.logStart) return

    console.log(
      [
        dimGrey(...Object.entries(opt.env || {}).map(([k, v]) => [k, v].join('='))),
        white(opt.name || _substringAfterLast(cmd, '/')),
        ...((!opt.name && (opt as SpawnOptions).args) || []),
      ]
        .filter(Boolean)
        .join(' '),
    )
  }

  private logFinish(
    cmd: string,
    opt: SpawnOptions | ExecOptions,
    started: UnixTimestampMillis,
    isSuccessful: boolean,
  ): void {
    if (isSuccessful && !opt.logFinish) return

    console.log(
      [
        white(opt.name || _substringAfterLast(cmd, '/')),
        ...((!opt.name && (opt as SpawnOptions).args) || []),
        dimGrey('took ' + _since(started)),
        !isSuccessful && dimGrey('and ') + dimRed('failed'),
      ]
        .filter(Boolean)
        .join(' '),
    )
  }
}

export const exec2 = new Exec2()

export class SpawnError extends AppError<SpawnErrorData> {
  constructor(message: string, data: SpawnErrorData) {
    super(message, data, { name: 'SpawnError' })
  }
}

export interface SpawnErrorData extends SpawnOutput {}

export interface SpawnOutput {
  /**
   * Exit code of the spawned process.
   * 0 means success, anything else means failure.
   */
  exitCode: number
  stdout: string
  stderr: string
}

export interface SpawnAsyncOptions extends SpawnOptions {
  /**
   * Defaults to true.
   * If true - prints both stdout and stderr to console while running,
   * otherwise runs "silently".
   * Returns SpawnOutput in the same way, regardless of `printWhileRunning` setting.
   */
  printWhileRunning?: boolean

  /**
   * Defaults to true.
   * If true - collects stdout and stderr while running, and return it in the end.
   * stdout/stderr are collected and returned regardless if it returns with error or not.
   * On success - stdout/stderr are available from `SpawnOutput`.
   * On error - stdout/stderr are available from `SpawnError.data`.
   */
  collectOutputWhileRunning?: boolean

  /**
   * Defaults to true.
   * If true - throws SpawnError if non-zero code is returned.
   * SpawnError conveniently contains .data.stdout and .data.strerr for inspection.
   * If false - will not throw, but return SpawnOutput with stdout, stderr and exitCode.
   */
  throwOnNonZeroCode?: boolean
}

export interface SpawnOptions {
  args?: string[]
  /**
   * Defaults to true.
   */
  logStart?: boolean
  /**
   * Defaults to true.
   */
  logFinish?: boolean
  /**
   * Defaults to true.
   * Controls/overrides both logStart and logFinish simultaneously.
   */
  log?: boolean
  /**
   * Defaults to true.
   */
  shell?: boolean

  /**
   * If specified - will be used as "command name" for logging purposes,
   * instead of "cmd + args"
   */
  name?: string
  cwd?: string

  env?: AnyObject
  /**
   * Defaults to true.
   * Set to false to NOT pass `process.env` to the spawned process.
   */
  passProcessEnv?: boolean
  /**
   * Defaults to "auto detect colors".
   * Set to false or true to override.
   */
  forceColor?: boolean
}

export interface ExecOptions {
  /**
   * Defaults to false.
   */
  logStart?: boolean
  /**
   * Defaults to false.
   */
  logFinish?: boolean
  /**
   * Defaults to false.
   * Controls/overrides both logStart and logFinish simultaneously.
   */
  log?: boolean

  /**
   * If specified - will be used as "command name" for logging purposes,
   * instead of "cmd + args"
   */
  name?: string
  cwd?: string
  timeout?: NumberOfMilliseconds

  env?: AnyObject
  /**
   * Defaults to false for security reasons.
   * Set to true to pass `process.env` to the spawned process.
   */
  passProcessEnv?: boolean
}
