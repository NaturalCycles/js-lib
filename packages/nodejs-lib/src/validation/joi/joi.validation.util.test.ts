import type { AnyObject } from '@naturalcycles/js-lib'
import { _stringify } from '@naturalcycles/js-lib'
import { expect, test } from 'vitest'
import { testValidation } from '../../test/validation.test.util.js'
import { Joi } from './joi.extensions.js'
import {
  arraySchema,
  booleanDefaultToFalseSchema,
  booleanSchema,
  emailSchema,
  integerSchema,
  numberSchema,
  objectSchema,
  stringSchema,
} from './joi.shared.schemas.js'
import { JoiValidationError } from './joi.validation.error.js'
import {
  convert,
  getValidationResult,
  isValid,
  undefinedIfInvalid,
  validate,
} from './joi.validation.util.js'

class Obj1 {
  a1!: string
  a2?: string
}

const obj1Schema = objectSchema<Obj1>({
  a1: stringSchema.min(2).max(5),
  a2: stringSchema.min(2).optional(),
})

interface Obj2 {
  s1?: string
  a: Obj1[]
}

const obj2Schema = objectSchema<Obj2>({
  s1: stringSchema.optional(),
  a: arraySchema<Obj1>(obj1Schema),
})

const invalidValues: any[] = [
  undefined,
  null,
  '',
  'abc',
  5,
  () => 'a',
  { a1: 'a' },
  { a1: 'a12345' },
]

const validValues: any[] = [
  {
    a1: 'ff',
  },
  {
    a1: 'ff',
    a2: 'sdfer',
  },
]

test('should fail on invalid values', () => {
  invalidValues.forEach(v => {
    expect(() => validate(v, obj1Schema)).toThrowErrorMatchingSnapshot()
  })
})

test('should pass on valid values', () => {
  validValues.forEach(v => {
    validate(v, obj1Schema)
  })
})

test('should trim strings by default', async () => {
  const v = { a1: ' sdf ' }
  const v2 = validate(v, obj1Schema)

  expect(v2.a1).toBe('sdf')
  expect(v === v2).toBe(false) // object should be cloned
})

test('should strip unknown keys', async () => {
  const v: AnyObject = {
    a1: 'ff',
    unk: 'ddd',
  }
  const v2 = validate(v, obj1Schema)

  expect(v2).toEqual({ a1: 'ff' })
  expect((v2 as any)['unk']).toBeUndefined()
})

test('getValidationResult should still convert', async () => {
  const v = {
    a1: ' ff ', // to be converted
    a2: 'a', // invalid!
  }
  const vr = getValidationResult(v, obj1Schema, 'objName')
  expect(vr.value.a1).toBe('ff')
  expect(vr.error).toBeInstanceOf(JoiValidationError)
  expect(vr.error).toMatchSnapshot()
})

test('getValidationResult valid', async () => {
  const vr = getValidationResult('asd', stringSchema)
  expect(vr.error).toBeUndefined()
})

test('error should contain errorItems', async () => {
  const v = {
    a1: ' ff ', // to be converted
    a2: 'a', // invalid!
  }
  const { error } = getValidationResult(v, obj1Schema, 'objName')
  expect(error!.data).toMatchSnapshot()
})

test('array items with unknown props', async () => {
  const v: Obj2 = {
    a: [
      {
        a1: 'hello',
        unk: 'unk', // unknown property!
      } as Obj1,
    ],
  }

  const { value, error } = getValidationResult(v, obj2Schema)
  // console.log(value)
  expect(error).toBeUndefined()

  // Expect 'unk' to be stripped away, no validation error
  expect(value).toEqual({
    a: [{ a1: 'hello' }],
  })
})

test('array with invalid items', async () => {
  const v: Obj2 = {
    a: [
      {
        a1: 'hello',
      } as Obj1,
      '' as any, // invalid value
    ],
  }

  const { error } = getValidationResult(v, obj2Schema)
  // console.log(value)
  expect(error!.data).toMatchSnapshot()
})

test('array items with invalid props', async () => {
  const v: Obj2 = {
    a: [
      {
        a1: 5, // invalid type
      } as any,
    ],
  }

  const { error } = getValidationResult(v, obj2Schema)
  expect(error!.data).toMatchSnapshot()
})

// optional
// default values

test('long message string', () => {
  const objSchema = arraySchema<Obj1>(obj1Schema)

  const longObject = Array.from({ length: 1000 }).fill({ a1: 5 })

  const { error } = getValidationResult(longObject, objSchema)
  // console.log(error!.message, error!.message.length)
  expect(error!.message).toMatchSnapshot()
})

test('should include id in the error message', () => {
  const obj = {
    id: 'someId',
  }

  const obj1 = Object.assign(new Obj1(), {
    id: 'someId',
  })

  // No objectName, with id
  let { error } = getValidationResult(obj as any, obj1Schema)
  // console.log(error)
  expect(error!.message).toMatchSnapshot()
  expect(error!.data).toMatchSnapshot()

  // ObjectName, with id
  ;({ error } = getValidationResult(obj as any, obj1Schema, 'ObjName'))
  // console.log(error)
  expect(error!.message).toMatchSnapshot()
  expect(error!.data).toMatchSnapshot()

  // No objectName, with id, constructor name
  ;({ error } = getValidationResult(obj1, obj1Schema))
  // console.log(error)
  expect(error!.message).toMatchSnapshot()
  expect(error!.data).toMatchSnapshot()
})

test('should return value on undefined schema', () => {
  const obj = {
    id: 'someId',
  }

  expect(validate(obj)).toBe(obj)
  const { value, error } = getValidationResult(obj)
  expect(value).toBe(obj)
  expect(error).toBeUndefined()
})

test('should convert empty string to undefined and strip', () => {
  const schema = objectSchema<Obj1>({
    a1: stringSchema.optional(),
    a2: stringSchema.optional(),
  })

  const obj1: Obj1 = {
    a1: 'asd ',
    a2: '',
  }

  const obj1mod = validate(obj1, schema)
  // console.log(obj1mod)
  expect(obj1mod).toEqual({
    a1: 'asd',
  })
})

test('empty string is not valid stringSchema', () => {
  expect(isValid('', stringSchema)).toBe(false)
})

test('should convert null to undefined and strip', () => {
  const schema = objectSchema<Obj1>({
    a1: stringSchema.optional(),
    // this is how `null` can be accepted (as empty):
    a2: stringSchema.empty(null).optional(),
  })

  const obj1: Obj1 = {
    a1: 'asd ',
    a2: null as any,
  }

  const obj1mod = validate(obj1, schema)
  // console.log(obj1mod)
  expect(obj1mod).toEqual({
    a1: 'asd',
  })

  // undefined is fine (empty, optional)
  expect(validate(undefined, booleanSchema.optional())).toBeUndefined()
  expect(validate(undefined, numberSchema.optional())).toBeUndefined()

  // null is invalid
  expect(isValid(null, numberSchema.optional())).toBe(false)
  expect(isValid(null, integerSchema.optional())).toBe(false)
  expect(isValid(null, arraySchema().optional())).toBe(false)

  // this is how to make null valid
  const numberPermissiveSchema = numberSchema.empty(null).optional()
  expect(isValid(null, numberPermissiveSchema)).toBe(true)
  expect(isValid(undefined, numberPermissiveSchema)).toBe(true)
  expect(isValid(5.1, numberPermissiveSchema)).toBe(true)
})

test('null is not a valid value when required', () => {
  testValidation(stringSchema, ['a'], ['', null, undefined, 5])
  testValidation(numberSchema, [5, 0], ['', null, undefined, 'a'])
  testValidation(
    booleanSchema,
    [true, false, 'true', 'false'],
    ['', null, undefined, 'a', 'tr', 'fa', 0, -1],
  )
})

test('default to empty array', () => {
  expect(validate(undefined, arraySchema().optional())).toBeUndefined()
  expect(validate(undefined, arraySchema().optional().default([]))).toEqual([])
  // expect(validate(null, arraySchema().optional().default([]))).toEqual([])
})

// Checking that partial schema is allowed (not all keys of Obj1 are required)
const _partialSchema = objectSchema<Obj1>({
  a2: stringSchema,
})

test('isValid', () => {
  expect(isValid('asd', stringSchema)).toBe(true)
  expect(isValid(56 as any, stringSchema)).toBe(false)
})

test('undefinedIfInvalid', () => {
  expect(undefinedIfInvalid('asd', stringSchema)).toBe('asd')
  expect(undefinedIfInvalid(56 as any, stringSchema)).toBeUndefined()
})

test('convert', () => {
  expect(convert(undefined)).toBeUndefined()
  expect(convert(undefined, stringSchema)).toBeUndefined()
  expect(convert('', stringSchema)).toBeUndefined()
  expect(convert('', stringSchema.optional())).toBeUndefined()
  expect(convert('a', stringSchema)).toBe('a')
  expect(convert(' a', stringSchema)).toBe('a')
  expect(convert(' a b  ', stringSchema)).toBe('a b')
})

test('booleanDefaultToFalseSchema', () => {
  const s = booleanDefaultToFalseSchema
  expect(convert(undefined, s)).toBe(false)
  expect(convert(false, s)).toBe(false)
  expect(convert(true, s)).toBe(true)
})

// todo
test.skip('arraySchema should strip undefined/null values by default', () => {
  // const schema = arraySchema(stringSchema)
  const schema = Joi.array().sparse(true).items(stringSchema.optional())

  expect(validate(['s', undefined, 's2', null], schema)).toEqual(['s', 's2'])
})

test('annotation is non-enumerable, but still accessible', () => {
  const s = booleanSchema
  const { error } = getValidationResult('notBoolean' as any, s)
  expect(error).toBeInstanceOf(JoiValidationError)

  expect(error!.data).toMatchInlineSnapshot(`
    {
      "joiValidationErrorItems": [
        {
          "context": {
            "label": "value",
            "value": "notBoolean",
          },
          "message": ""value" must be a boolean",
          "path": [],
          "type": "boolean.base",
        },
      ],
    }
  `)

  expect(JSON.stringify(error!.data)).not.toContain('annotation')

  // But it's still present!
  expect(error!.data.annotation).toMatchInlineSnapshot(`""value" must be a boolean"`)
  expect(!!error!.data.annotation).toBe(true)
})

test('formatting of email error', () => {
  const obj = {
    name1: 'Vasya',
    name2: 'Pupkeen',
    email: 'la@gmail.con',
  }

  const schema = objectSchema({
    name1: stringSchema,
    name2: stringSchema,
    email: emailSchema,
  })

  const { error } = getValidationResult(obj, schema)
  expect(_stringify(error)).toMatchInlineSnapshot(`
    "JoiValidationError: "email" must be a valid email

    {
      "name1": "Vasya",
      "name2": "Pupkeen",
      "email" [1]: "la@gmail.con"
    }

    [1] "email" must be a valid email"
  `)
})
