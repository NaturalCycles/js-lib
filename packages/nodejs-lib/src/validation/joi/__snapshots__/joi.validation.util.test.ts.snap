// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`array items with invalid props 1`] = `
{
  "joiValidationErrorItems": [
    {
      "context": {
        "key": "a1",
        "label": "a[0].a1",
        "value": 5,
      },
      "message": ""a[0].a1" must be a string",
      "path": [
        "a",
        0,
        "a1",
      ],
      "type": "string.base",
    },
  ],
}
`;

exports[`array with invalid items 1`] = `
{
  "joiValidationErrorItems": [
    {
      "context": {
        "key": 1,
        "label": "a[1]",
        "type": "object",
        "value": "",
      },
      "message": ""a[1]" must be of type object",
      "path": [
        "a",
        1,
      ],
      "type": "object.base",
    },
  ],
}
`;

exports[`error should contain errorItems 1`] = `
{
  "joiValidationErrorItems": [
    {
      "context": {
        "encoding": undefined,
        "key": "a2",
        "label": "a2",
        "limit": 2,
        "value": "a",
      },
      "message": ""a2" length must be at least 2 characters long",
      "path": [
        "a2",
      ],
      "type": "string.min",
    },
  ],
  "joiValidationObjectName": "objName",
}
`;

exports[`getValidationResult should still convert 1`] = `
[JoiValidationError: Invalid objName
{
  "a1": " ff ",
  "a2" [1]: "a"
}

[1] "a2" length must be at least 2 characters long]
`;

exports[`long message string 1`] = `
""[0].a1" must be a string
"[1].a1" must be a string
"[2].a1" must be a string
"[3].a1" must be a string
"[4].a1" must be a string
... 1000 errors in total

[
  {
    "a1" [1, 10, 100, 1000, 101, 102, 103, 104, 105, 106, 107, 108, 109, 11, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 12, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 13, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 14, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 15, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 16, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 17, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 18, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 19, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 2, 20, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 21, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219, 22, 220, 221, 222, 223, 224, 225, 226, 227, 228, 229, 23, 230, 231, 232, 233, 234, 235, 236, 237, 238, 239, 24, 240, 241, 242, 243, 244, 245, 246, 247, 248, 249, 25, 250, 251, 252, 253, 254, 255, 256, 257, 258, 259, 26, 260, 261, 262, 263, 264, 265, 266, 267, 268, 269, 27, 270, 271, 272, 273, 274, 275, 276, 277, 278, 279, 28, 280, 281, 282, 283, 284, 285, 286, 287, 288, 289, 29, 290, 291, 292, 293, 294, 295, 296, 297, 298, 299, 3, 30, 300, 301, 302, 303, 304, 305, 306, 307, 308, 309, 31, 310, 311, 312, 313, 314, 315, 316, 317, 318, 319, 32, 320, 321, 322, 323, 324, 325, 326, 327, 328, 329, 33, 330, 331, 332, 333, 334, 335, 336, 337, 338, 339, 34, 340, 341, 342, 343, 344, 345, 346, 347, 348, 349, 35, 350, 351, 352, 353, 354, 355, 356, 357, 358, 359, 36, 360, 361, 362, 363, 364, 365, 366, 367, 368, 369, 37, 370, 371, 372, 373, 374, 375, 376, 377, 378, 379, 38, 380, 381, 382, 383, 384, 385, 386, 387, 388, 389, 39, 390, 391, 392, 393, 394, 395, 396, 397, 398, 399, 4, 40, 400, 401, 402, 403, 404, 405, 406, 407, 408, 409, 41, 410, 411, 412, 413, 414, 415, 416, 417, 418, 419, 42, 420, 421, 422, 423, 424, 425, 426, 427, 428, 429, 43, 430, 431, 432, 433, 434, 435, 436, 437, 438, 439, 44, 440, 441, 442, 443, 444, 445, 446, 447, 448, 449, 45, 450, 451, 452, 453, 454, 455, 456, 457, 458, 459, 46, 4
... 43.6 Kb message truncated ...
 a string
[943] "[942].a1" must be a string
[944] "[943].a1" must be a string
[945] "[944].a1" must be a string
[946] "[945].a1" must be a string
[947] "[946].a1" must be a string
[948] "[947].a1" must be a string
[949] "[948].a1" must be a string
[950] "[949].a1" must be a string
[951] "[950].a1" must be a string
[952] "[951].a1" must be a string
[953] "[952].a1" must be a string
[954] "[953].a1" must be a string
[955] "[954].a1" must be a string
[956] "[955].a1" must be a string
[957] "[956].a1" must be a string
[958] "[957].a1" must be a string
[959] "[958].a1" must be a string
[960] "[959].a1" must be a string
[961] "[960].a1" must be a string
[962] "[961].a1" must be a string
[963] "[962].a1" must be a string
[964] "[963].a1" must be a string
[965] "[964].a1" must be a string
[966] "[965].a1" must be a string
[967] "[966].a1" must be a string
[968] "[967].a1" must be a string
[969] "[968].a1" must be a string
[970] "[969].a1" must be a string
[971] "[970].a1" must be a string
[972] "[971].a1" must be a string
[973] "[972].a1" must be a string
[974] "[973].a1" must be a string
[975] "[974].a1" must be a string
[976] "[975].a1" must be a string
[977] "[976].a1" must be a string
[978] "[977].a1" must be a string
[979] "[978].a1" must be a string
[980] "[979].a1" must be a string
[981] "[980].a1" must be a string
[982] "[981].a1" must be a string
[983] "[982].a1" must be a string
[984] "[983].a1" must be a string
[985] "[984].a1" must be a string
[986] "[985].a1" must be a string
[987] "[986].a1" must be a string
[988] "[987].a1" must be a string
[989] "[988].a1" must be a string
[990] "[989].a1" must be a string
[991] "[990].a1" must be a string
[992] "[991].a1" must be a string
[993] "[992].a1" must be a string
[994] "[993].a1" must be a string
[995] "[994].a1" must be a string
[996] "[995].a1" must be a string
[997] "[996].a1" must be a string
[998] "[997].a1" must be a string
[999] "[998].a1" must be a string
[1000] "[999].a1" must be a string"
`;

exports[`should fail on invalid values 1`] = `[JoiValidationError: "value" is required]`;

exports[`should fail on invalid values 2`] = `[JoiValidationError: "value" must be of type object]`;

exports[`should fail on invalid values 3`] = `[JoiValidationError: "value" must be of type object]`;

exports[`should fail on invalid values 4`] = `[JoiValidationError: "value" must be of type object]`;

exports[`should fail on invalid values 5`] = `[JoiValidationError: "value" must be of type object]`;

exports[`should fail on invalid values 6`] = `[JoiValidationError: "value" must be of type object]`;

exports[`should fail on invalid values 7`] = `
[JoiValidationError: {
  "a1" [1]: "a"
}

[1] "a1" length must be at least 2 characters long]
`;

exports[`should fail on invalid values 8`] = `
[JoiValidationError: {
  "a1" [1]: "a12345"
}

[1] "a1" length must be less than or equal to 5 characters long]
`;

exports[`should include id in the error message 1`] = `
"Invalid Object.someId
{
  "id": "someId",
  "a1" [1]: -- missing --
}

[1] "a1" is required"
`;

exports[`should include id in the error message 2`] = `
{
  "joiValidationErrorItems": [
    {
      "context": {
        "key": "a1",
        "label": "a1",
      },
      "message": ""a1" is required",
      "path": [
        "a1",
      ],
      "type": "any.required",
    },
  ],
  "joiValidationObjectId": "someId",
  "joiValidationObjectName": "Object",
}
`;

exports[`should include id in the error message 3`] = `
"Invalid ObjName.someId
{
  "id": "someId",
  "a1" [1]: -- missing --
}

[1] "a1" is required"
`;

exports[`should include id in the error message 4`] = `
{
  "joiValidationErrorItems": [
    {
      "context": {
        "key": "a1",
        "label": "a1",
      },
      "message": ""a1" is required",
      "path": [
        "a1",
      ],
      "type": "any.required",
    },
  ],
  "joiValidationObjectId": "someId",
  "joiValidationObjectName": "ObjName",
}
`;

exports[`should include id in the error message 5`] = `
"Invalid Obj1.someId
{
  "id": "someId",
  "a1" [1]: -- missing --
}

[1] "a1" is required"
`;

exports[`should include id in the error message 6`] = `
{
  "joiValidationErrorItems": [
    {
      "context": {
        "key": "a1",
        "label": "a1",
      },
      "message": ""a1" is required",
      "path": [
        "a1",
      ],
      "type": "any.required",
    },
  ],
  "joiValidationObjectId": "someId",
  "joiValidationObjectName": "Obj1",
}
`;
